src/ngx_lua_resty_lmdb_module.c

```diff
@@ -289,26 +290,136 @@ static ngx_int_t ngx_lua_resty_lmdb_init_worker(ngx_cycle_t *cycle)
+            ngx_explicit_memzero(lcf->encryption_mode.data, lcf->encryption_mode.len);
+
+            ngx_str_null(&lcf->key_data);
+            ngx_str_null(&lcf->encryption_mode);
+        }
+    }
+
+    return NGX_OK;
+}
+
+
+static ngx_int_t
+ngx_lua_resty_lmdb_remove_files(ngx_cycle_t *cycle, ngx_path_t *path)
+{
+    ngx_file_info_t   fi;
+    ngx_err_t         err;
+
+    u_char            name_buf[NGX_MAX_PATH];
+    ngx_str_t        *names = LMDB_FILE_NAMES + 1;
+    ngx_str_t        *name;
+
+    if (ngx_file_info(path->name.data, &fi) == NGX_FILE_ERROR) {
+        ngx_log_error(NGX_LOG_CRIT, cycle->log, ngx_errno,
+                      ngx_file_info_n " \"%s\" failed", path->name.data);
+        return NGX_ERROR;
+    }
+
+    if (ngx_is_dir(&fi)) {
+
+        /* try to remove all lmdb files */
+        for (name = names; name->len; name++) {
+            ngx_snprintf(name_buf, NGX_MAX_PATH,
+                         "%V%V%Z", &path->name, name);
+
+            ngx_log_debug1(NGX_LOG_DEBUG_CORE, cycle->log, 0,
+                           "lmdb file remove: \"%s\"", name_buf);
+
+            if (ngx_delete_file(name_buf) == NGX_FILE_ERROR) {
+                ngx_log_error(NGX_LOG_WARN, cycle->log, ngx_errno,
+                              ngx_delete_file_n " \"%s\" failed",
+                              name_buf);
+            }
+        }
+
+        return NGX_OK;
+    }
+
+    ngx_lua_resty_lmdb_assert(!ngx_is_dir(&fi));
+
+    /* try to delete the file */
+    if (ngx_delete_file(path->name.data) == NGX_FILE_ERROR) {
+        ngx_log_error(NGX_LOG_WARN, cycle->log, ngx_errno,
+                      ngx_delete_file_n " \"%V\" failed",
+                      &path->name);
+    }
+
+    /* add a slash at the tail */
+    if (path->name.data[path->name.len - 1] == '/') {
+        ngx_snprintf(name_buf, NGX_MAX_PATH, "%V%Z", &path->name);
+
+    } else {
+        ngx_snprintf(name_buf, NGX_MAX_PATH, "%V/%Z", &path->name);
+    }
+
+    /* ensure lmdb directory exists */
+    err = ngx_create_full_path(name_buf, LMDB_DIR_MODE);
+    if (err) {
+        ngx_log_error(NGX_LOG_WARN, cycle->log, err,
+                      ngx_create_dir_n " \"%s\" failed", name_buf);
+    }
+
+    return NGX_OK;
+}
+
+
+static ngx_int_t
+ngx_lua_resty_lmdb_open_file(ngx_cycle_t *cycle,
+                             ngx_lua_resty_lmdb_conf_t *lcf,
+                             ngx_flag_t is_master)
+{
+    int                        rc;
+    int                        dead;
+    int                        try_count = 1;
+
+retry:
+
+    if (ngx_lua_resty_lmdb_create_env(cycle, lcf, is_master) != NGX_OK) {
+        return NGX_ERROR;
+    }
+
+    rc = mdb_env_open(lcf->env, (const char *) lcf->env_path->name.data,
+                      0, LMDB_FILE_MODE);
 
-        ngx_str_null(&lcf->key_data);
-        ngx_str_null(&lcf->encryption_mode);
+    if (rc == 0 || is_master == 0 || try_count == 0) {
```

Comment:
It seems that this is not 100% safe. If `rc != 0` on the second try, then we can not safely do `check_lmdb` because environment is not even open.

We don't need this `goto` logic, suggest we just check for:

`MDB_VERSION_MISMATCH || MDB_INVALID`

and in these case, remove the file and `mdb_env_open` again. For all other errors, we should just fail because file removal will not solve the issue.

Reply:
done, check `ENOTDIR`.